backend/danswer/datastores/vespa/store.py

```
@@ -0,0 +1,402 @@
+    }
+    while True:
+        results = requests.get(SEARCH_ENDPOINT, params=params).json()
+        hits = results["root"].get("children", [])
+        doc_chunk_ids.extend(
+            [hit.get("fields", {}).get("documentid").split("::")[1] for hit in hits]
+        )
+        params["offset"] += hits_per_page  # type: ignore
+
+        if len(hits) < hits_per_page:
+            break
+    return doc_chunk_ids
+
+
+def _delete_vespa_doc_chunks(document_id: str) -> bool:
+    doc_chunk_ids = _get_vespa_chunk_ids_by_document_id(document_id)
+
+    failures = [
+        requests.delete(f"{DOCUMENT_ID_ENDPOINT}/{doc}").status_code != 200
+        for doc in doc_chunk_ids
+    ]
+    return not any(failures)
+
+
+def _index_vespa_chunks(
+    chunks: list[IndexChunk],
+    index_attempt_metadata: IndexAttemptMetadata,
+) -> set[DocumentInsertionRecord]:
+    json_header = {"Content-Type": "application/json"}
+    insertion_records: set[DocumentInsertionRecord] = set()
+    cross_connector_document_metadata_map: dict[
+        str, CrossConnectorDocumentMetadata
+    ] = {}
+    # document ids of documents that existed BEFORE this indexing
+    already_existing_documents: set[str] = set()
+    for chunk in chunks:
+        document = chunk.source_document
+        (
+            cross_connector_document_metadata_map,
+            should_delete_doc,
+        ) = update_cross_connector_document_metadata_map(
+            chunk=chunk,
+            cross_connector_document_metadata_map=cross_connector_document_metadata_map,
+            doc_store_cross_connector_document_metadata_fetch_fn=partial(
+                _get_vespa_document_cross_connector_metadata,
+            ),
+            index_attempt_metadata=index_attempt_metadata,
+        )
+
+        if should_delete_doc:
+            # Processing the first chunk of the doc and the doc exists
+            deletion_success = _delete_vespa_doc_chunks(document.id)
+            if not deletion_success:
```

nit: prefer to fail loudly in case that we fail to delete (plus add retries to delete)